File: Java/src/main/java/UnsupervisedCompetativeNeuralNet.java
import java.util.Random;

/**
 * Created by Daniel on 19/05/2015.
 */
public class UnsupervisedCompetativeNeuralNet {
    private final int inputNodesCounts;
    private final int hiddenNodesCount;
    private float learningRate;
    private float biasLearningRate;
    private final float[][] connections;
    private final float[] biases;
    private final float[] hiddenNodes;

    public UnsupervisedCompetativeNeuralNet(int inputNodesCounts, int hiddenNodesCount,
                                            float learningRate, float biasLearningRate){
        this.inputNodesCounts = inputNodesCounts;
        this.hiddenNodesCount = hiddenNodesCount;
        this.learningRate = learningRate;
        this.biasLearningRate = biasLearningRate;
        biases = new float[this.hiddenNodesCount];
        hiddenNodes = new float[this.hiddenNodesCount];

        Random random = new Random();
        connections = new float[this.inputNodesCounts][];
        for (int i = 0; i < inputNodesCounts; i++){
            connections[i] = new float[this.hiddenNodesCount];
            for (int h =0; h < hiddenNodesCount; h++){
                connections[i][h] = random.nextFloat();
            }
        }
    }

    private void FeedForward(float[] data){
        for(int h=0; h<hiddenNodesCount;h++){
            float activation = biases[h];
            for(int i=0; i<inputNodesCounts; i++){
                activation += data[i]*connections[i][h];
            }

            this.hiddenNodes[h] = activation;
        }
    }

    public int train(float[] data){
        FeedForward(data);
        int winningNode = getIndexOfMax(hiddenNodes);

        for(int i = 0; i< inputNodesCounts; i++){
            connections[i][winningNode] += learningRate*(data[i]-connections[i][winningNode]);
        }

        return winningNode;
    }

    public void train(Iterable<float[]> dataSet){
        for (float[] data : dataSet) {
            train(data);
        }
    }

    public int getCluster(float[] data){
        FeedForward(data);
        return getIndexOfMax(this.hiddenNodes);
    }

    private static int getIndexOfMax(float[] items){
        float maxValue = items[0];
        int maxIndex = 0;
        for(int i=1;i<items.length;i++){
            if(items[i] > maxValue){
                maxValue = items[i];
                maxIndex = i;
            }
        }

        return maxIndex;
    }

    public float getLearningRate() {
        return learningRate;
    }

    public void setLearningRate(float learningRate) {
        this.learningRate = learningRate;
    }

    public float getBiasLearningRate() {
        return biasLearningRate;
    }

    public void setBiasLearningRate(float biasLearningRate) {
        this.biasLearningRate = biasLearningRate;
    }
}

File: C#/UnsupervisedCompetativeNeuralNet/UnsupervisedCompetativeNeuralNet.cs
using System;
using System.Collections.Generic;

namespace UnsupervisedCompetativeNeuralNet
{
    public class UnsupervisedCompetativeNeuralNet
    {
        private readonly float[] biases;
        private readonly float[][] connections;
        private readonly float[] hiddenNodes;
        private readonly int hiddenNodesCount;
        private readonly int inputNodesCount;

        public UnsupervisedCompetativeNeuralNet(int inputNodesCount, int hiddenNodesCount, float learningRate,
            float biasLearningRate)
        {
            if (hiddenNodesCount < 1)
                throw new ArgumentException("must be greater than 1", "hiddenNodesCount");
            if (inputNodesCount < 0)
                throw new ArgumentException("must be greater than 0", "inputNodesCount");
            if (learningRate < 0)
                throw new ArgumentException("must be greater than 0", "learningRate");
            if (learningRate > 1.0)
                throw new ArgumentException("must be less than 1", "learningRate");

            this.inputNodesCount = inputNodesCount;
            this.hiddenNodesCount = hiddenNodesCount;
            LearningRate = learningRate;
            BiasLearningRate = biasLearningRate;
            biases = new float[this.hiddenNodesCount];
            connections = new float[this.inputNodesCount][];

            var random = new Random();
            for (var i = 0; i < this.inputNodesCount; i++)
            {
                connections[i] = new float[this.hiddenNodesCount];
                for (var j = 0; j < this.hiddenNodesCount; j++)
                    connections[i][j] = (float) random.NextDouble();
            }
            hiddenNodes = new float[this.hiddenNodesCount];
        }

        public float LearningRate { get; set; }
        public float BiasLearningRate { get; set; }

        public int GetCluster(float[] data)
        {
            FeedForward(data);
            return GetIndexOfMax(hiddenNodes);
        }

        public int Train(float[] data)
        {
            FeedForward(data);
            var winningCluster = GetIndexOfMax(hiddenNodes);
            for (var i = 0; i < inputNodesCount; i++)
                connections[i][winningCluster] += LearningRate*(data[i] - connections[i][winningCluster]);

            biases[winningCluster] -= BiasLearningRate;

            return winningCluster;
        }

        public void Train(IEnumerable<float[]> dataSet)
        {
            foreach (var data in dataSet)
                Train(data);
        }

        private static int GetIndexOfMax(float[] array)
        {
            var maxIndex = 0;
            var maxValue = array[0];
            for (var i = 1; i < array.Length; i++)
            {
                if (array[i] > maxValue)
                {
                    maxValue = array[i];
                    maxIndex = i;
                }
            }

            return maxIndex;
        }

        private void FeedForward(float[] data)
        {
            for (var i = 0; i < hiddenNodesCount; i++)
            {
                var activation = biases[i];
                for (var j = 0; j < inputNodesCount; j++)
                    activation += data[j] - connections[j][i];
                hiddenNodes[i] = activation;
            }
        }
    }
}